sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -642,15 +642,17 @@ object CombineConcats extends Rule[LogicalPlan] {
   private def flattenConcats(concat: Concat): Concat = {
     val stack = Stack[Expression](concat)
     val flattened = ArrayBuffer.empty[Expression]
+    var isBinaryMode = concat.isBinaryMode
     while (stack.nonEmpty) {
       stack.pop() match {
-        case Concat(children) =>
+        case Concat(children, binary) =>
+          isBinaryMode &= binary
```

Comment:
We need to update the comment of this optimizer rule to explain how `typeCoercionRules ` works for these nested cases.